Some more documentation progress (#345)
Jutho/TensorKit.jl: docs/src/man/tensormanipulations.md

of tensor factorizations for both endomorphisms and general homomorphism that we discuss
below.

Finally, there are a number of operations that also belong in this paragraph because of
their analogy to common matrix operations. The tensor product of two `TensorMap` instances
`t1` and `t2` is obtained as `t1 ⊗ t2` and results in a new `TensorMap` with
`codomain(t1 ⊗ t2) = codomain(t1) ⊗ codomain(t2)` and
`domain(t1 ⊗ t2) = domain(t1) ⊗ domain(t2)`. If we have two `TensorMap{T,S,N,1}` instances
`t1` and `t2` with the same codomain, we can combine them in a way that is analoguous to
`hcat`, i.e. we stack them such that the new tensor `catdomain(t1, t2)` has also the same
codomain, but has a domain which is `domain(t1) ⊕ domain(t2)`. Similarly, if `t1` and `t2`
are of type `TensorMap{T,S,1,N}` and have the same domain, the operation
`catcodomain(t1, t2)` results in a new tensor with the same domain and a codomain given by
`codomain(t1) ⊕ codomain(t2)`, which is the analogy of `vcat`. Note that direct sum only
makes sense between `ElementarySpace` objects, i.e. there is no way to give a tensor product
meaning to a direct sum of tensor product spaces.

Time for some more examples:

> ```@repl tensors

Reviewer: This REPL is failing because TensorKit isn't exported, and `t` and `V1` aren't defined

of which are reexported by TensorKit.jl. Note that `transpose(t)` is not simply equal to
reshuffling domain and codomain with
`braid(t, (1:(N₁+N₂)...), reverse(domainind(tsrc)), reverse(codomainind(tsrc))))`. Indeed,
the graphical representation (where we draw the codomain and domain as a single object),
makes clear that this introduces an additional (inverse) twist, which is then compensated in
the `transpose` implementation.

```@raw html
<img src="../img/tensor-transpose.svg" alt="transpose" class="color-invertible"/>
```

In categorical language, the reason for this extra twist is that we use the left
coevaluation ``η``, but the right evaluation ``\tilde{ϵ}``, when repartitioning the indices
between domain and codomain.

There are a number of other index related manipulations. We can apply a twist (or inverse
twist) to one of the tensor map indices via [`twist(t, i; inv = false)`](@ref) or
[`twist!(t, i; inv = false)`](@ref). Note that the latter method does not store the result
in a new destination tensor, but just modifies the tensor `t` in place. Twisting several
indices simultaneously can be obtained by using the defining property

> ``θ_{V⊗W} = τ_{W,V} ∘ (θ_W ⊗ θ_V) ∘ τ_{V,W} = (θ_V ⊗ θ_W) ∘ τ_{W,V} ∘ τ_{V,W}.``

Reviewer: ```suggestion
``θ_{V⊗W} = τ_{W,V} ∘ (θ_W ⊗ θ_V) ∘ τ_{V,W} = (θ_V ⊗ θ_W) ∘ τ_{W,V} ∘ τ_{V,W},``
```

but is currently not implemented explicitly.

For all sector types `I` with `BraidingStyle(I) == Bosonic()`, all twists are `1` and thus
have no effect. Let us start with some examples, in which we illustrate that, albeit
`permute` might act highly non-trivial on the fusion trees and on the corresponding data,
after conversion to a regular `Array` (when possible), it just acts like `permutedims`
```@repl tensors
domain(t) → codomain(t)
ta = convert(Array, t);
t′ = permute(t, (1,2,3,4));
domain(t′) → codomain(t′)
convert(Array, t′) ≈ ta
t′′ = permute(t, (4,2,3),(1,));
domain(t′′) → codomain(t′′)
convert(Array, t′′) ≈ permutedims(ta, (4,2,3,1))
m
transpose(m)
convert(Array, transpose(t)) ≈ permutedims(ta,(4,3,2,1))
dot(t2, t) ≈ dot(transpose(t2), transpose(t))
transpose(transpose(t)) ≈ t
twist(t, 3) ≈ t
# as twist acts trivially for
BraidingStyle(sectortype(t))
```
Note that `transpose` acts like one would expect on a `TensorMap{T,S,1,1}`. On a

> `TensorMap{TS,N₁,N₂}`, because `transpose` replaces the codomain with the dual of the

Reviewer: ```suggestion
`TensorMap{T,S,N₁,N₂}`, because `transpose` replaces the codomain with the dual of the
```

domain, which has its tensor product operation reversed, this in the end amounts in a
complete reversal of all tensor indices when representing it as a plain mutli-dimensional
`Array`. Also, note that we have not defined the conjugation of `TensorMap` instances. One
definition that one could think of is `conj(t) = adjoint(transpose(t))`. However note that
`codomain(adjoint(tranpose(t))) == domain(transpose(t)) == dual(codomain(t))` and similarly
`domain(adjoint(tranpose(t))) == dual(domain(t))`, where `dual` of a `ProductSpace` is
composed of the dual of the `ElementarySpace` instances, in reverse order of tensor product.
This might be very confusing, and as such we leave tensor conjugation undefined. However,
note that we have a conjugation syntax within the context of [tensor contractions](@ref
ss_tensor_contraction).

To show the effect of `twist`, we now consider a type of sector `I` for which

> `BraidingStyle{I} != Bosonic()`. In particular, we use `FibonacciAnyon`. We cannot convert

Reviewer: ```suggestion
`BraidingStyle(I) != Bosonic()`. In particular, we use `FibonacciAnyon`. We cannot convert
```

the resulting `TensorMap` to an `Array`, so we have to rely on indirect tests to verify our
results.

> ```@repl tensors

Reviewer: Same here
